model/mqo/mqo_loader.py

```diff
@@ -60,111 +91,139 @@ def mirror_copy(self, vertex, xi):
+    def __eq__(self, other):
+        return self.color == other.color and self.tex_name == other.tex_name
 
 
 class Obj(object):
-
     def __init__(self):
         self.vertex = []
         self.faces  = []
         self.mirror = None
         self.mirror_axis = None
 
-    def __iadd__(self, othr):
-        vmap = [0]*len(othr.vertex)
-        n = len(self.vertex)
-        vs = []
-        for i, v in enumerate(othr.vertex):
+    def __iadd__(self, other):
+        """
+        Override the `+=` operator
+
+        WARNING: this mutates the other object by changing the indices in the
+        faces! It's possible that the other Obj we're appending will contain
+        invalid state after this function returns.
+        """
+        # vmap is a map from indices in other.vertex to indices in self.vertex.
+        # We create it when appending vertices from other to self, and then use
+        # it when appending faces from other to self.
+        vmap = [0] * len(other.vertex)
+        n = len(self.vertex)  # Index of the next vertex to add to self.vertex
+        for i, v in enumerate(other.vertex):
             try:
                 vmap[i] = self.vertex.index(v)
             except ValueError:
                 vmap[i] = n
                 n += 1
-                vs.append(v)
-        self.vertex.extend(vs)
+                self.vertex.append(v)
 
-        fs = []
-        for face in othr.faces:
+        for face in other.faces:
             face.indices = [vmap[i] for i in face.indices]
             face.make_hash()
             if face not in self.faces:
-                fs.append(face)
-        self.faces.extend(fs)
+                self.faces.append(face)
         return self
 
     def expand_mirror(self):
-        if self.mirror not in [None, 0] and self.mirror_axis is not None:
-            x = y = z = 1.0
-            if self.mirror_axis&1: x = -1.0
-            if self.mirror_axis&2: y = -1.0
-            if self.mirror_axis&4: z = -1.0
-            xi = [x, y, z]
-            fs = []
-            for f in self.faces:
-                ff, c = f.mirror_copy(self.vertex, xi)
-                if c and ff in self.faces:
-                    self.vertex = self.vertex[:-c]
-                else:
-                    fs.append(ff)
-            self.faces.extend(fs)
-            self.mirror = None
+        """
+        For each face in the object, also include the face reflected across the
+        mirror, then remove the ability to mirror any more.
+
+        TODO: self.mirror appears to be unused. Can we use only mirror_axis
+        instead?
+        """
+        if self.mirror in [None, 0] or self.mirror_axis is None:
+            return
+
+        x = y = z = 1.0
+        if self.mirror_axis & 1: x = -1.0
+        if self.mirror_axis & 2: y = -1.0
+        if self.mirror_axis & 4: z = -1.0
+        xi = [x, y, z]
+        self.faces.expand(f.create_mirror(self.vertex, xi) for f in self.faces)
```

Comment:
Should be `self.faces.extend`. but that also implies that this code is unused in the game, because the game still runs normally.